bundles/org.openhab.binding.folderwatcher/src/main/java/org/openhab/binding/folderwatcher/internal/handler/FolderWatcherHandler.java

```diff
@@ -0,0 +1,263 @@
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.openhab.binding.folderwatcher.internal.common.WatcherCommon;
+import org.openhab.binding.folderwatcher.internal.config.FolderWatcherConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link FolderWatcherHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Alexandr Salamatov - Initial contribution
+ */
+@NonNullByDefault
+public class FolderWatcherHandler extends BaseThingHandler {
+    private final Logger logger = LoggerFactory.getLogger(FolderWatcherHandler.class);
+    private @Nullable FolderWatcherConfiguration config;
+    private @Nullable File currentFtpListingFile;
+    private @Nullable ScheduledFuture<?> executionJob, initJob;
+    private @Nullable FTPClient ftp; // = new FTPClient();
+    private List<String> currentFtpListing = new ArrayList<>();
+    private List<String> previousFtpListing = new ArrayList<>();
+
+    public FolderWatcherHandler(Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+    }
+
+    @Override
+    public void initialize() {
+        config = getConfigAs(FolderWatcherConfiguration.class);
+        updateStatus(ThingStatus.UNKNOWN);
+        if (config.connectionTimeout <= 0) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "Connection timeout can't be negative");
+            return;
+        }
+        if (config.ftpPort < 0) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "FTP port can't be negative");
+            return;
+        }
+        if (config.ftpDir == null) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "FTP directory can't be empty");
+            return;
+        }
+
+        currentFtpListingFile = new File(ConfigConstants.getUserDataFolder() + File.separator + "FolderWatcher"
+                + File.separator + thing.getUID().getAsString().replace(':', '_') + ".data");
+        try {
+            previousFtpListing = WatcherCommon.initStorage(currentFtpListingFile);
+        } catch (IOException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+            logger.debug("Can't write file {}, error message {}", currentFtpListingFile, e.getMessage());
+            return;
+        }
+
+        if (config.pollInterval > 0) {
+            initJob = scheduler.scheduleWithFixedDelay(this::connectionKeepAlive, 0, config.pollInterval,
+                    TimeUnit.SECONDS);
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "Polling interval can't be null or negative");
```

Comment:
This configuration error should have been checked for before you initialized storage.